python/ray/tune/tests/test_tuner_restore.py

```diff
@@ -497,6 +508,141 @@ def load_checkpoint(self, checkpoint_path):
             assert result.metrics["score"] == 2
 
 
+def test_restore_overwrite_trainable(ray_start_4_cpus, tmpdir):
+    """Test validation for trainable compatibility, when re-specifying a trainable
+    on restore."""
+
+    def train_func_1(config):
+        data = {"data": config["data"]}
+        session.report(data, checkpoint=Checkpoint.from_dict(data))
+        raise RuntimeError("Failing!")
+
+    tuner = Tuner(
+        train_func_1,
+        run_config=RunConfig(name="overwrite_trainable", local_dir=str(tmpdir)),
+        param_space={"data": 1},
+    )
+    tuner.fit()
+
+    del tuner
+    ray.shutdown()
+    ray.init(num_cpus=4)
+
+    # Can't overwrite with a different Trainable type
+    with pytest.raises(ValueError):
+        tuner = Tuner.restore(
+            str(tmpdir / "overwrite_trainable"),
+            overwrite_trainable="__fake",
+            resume_errored=True,
+        )
+
+    # Can't overwrite with a different Trainable name
+    def train_func_2(config):
+        checkpoint = session.get_checkpoint()
+        assert checkpoint and checkpoint.to_dict()["data"] == config["data"]
+
+    with pytest.raises(ValueError):
+        tuner = Tuner.restore(
+            str(tmpdir / "overwrite_trainable"),
+            overwrite_trainable=train_func_2,
+            resume_errored=True,
+        )
+
+    # Can still change trainable code, but logs a warning
+    def train_func_1(config):
+        checkpoint = session.get_checkpoint()
+        assert checkpoint and checkpoint.to_dict()["data"] == config["data"]
+
+    tuner = Tuner.restore(
```

Comment:
can we test that the warning is logged? https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#warns

Reply:
Should I warn using `logger.warning` or `warnings.warn`? Currently choosing to log it as a warning, since there's not always something actionable for the user to do.

Reply:
Ah got it- you're right `logger.warning` is what we want then I think. We can still check for this via the caplog fixtureL

```
def test(caplog):
    with caplog.set_level(logging.WARNING):
        Tuner.retore(...)
    assert ... in caplog.text
```

https://stackoverflow.com/questions/53125305/testing-logging-output-with-pytest
https://docs.pytest.org/en/latest/how-to/logging.html#caplog-fixture